junebug/logging_service.py

```diff
@@ -0,0 +1,127 @@
+        return self.DEFAULT_LOG_LEVEL
+
+    def logger_for_event(self, event):
+        '''Get the name of the logger for an event.'''
+        system = event.get('system', '-')
+        parts = [self.worker_id]
+        if system != '-':
+            parts.extend(system.split(','))
+        logger = ".".join(parts)
+        return logger.lower()
+
+    def _log_to_file(self, event):
+        '''Logs the specified event to the log file.'''
+        level = self.level_for_event(event)
+        if level < self.LOG_LEVEL_THRESHOLD:
+            return
+
+        data = {
+            "logger": self.logger_for_event(event),
+            "level": level,
+            "timestamp": time.time(),
+        }
+
+        failure = event.get('failure')
+        if failure:
+            data['class'] = repr(failure.type)
+            data['instance'] = repr(failure.value)
+            data['stack'] = failure.stack
+
+        data['message'] = log.textFromEventDict(event)
+
+        self.logfile.write(json.dumps(data))
+
+    def __call__(self, event):
+        if self.log_context_sentinel in event:
+            return
+        log.callWithContext(self.log_context, self._log_to_file, event)
+
+
+class JunebugLoggerService(Service):
+    '''Service for :class:`junebug.logging.JunebugLogObserver`'''
+
+    def __init__(self, worker_id, path, rotate, max_files, logger=None):
+        '''
+        Create the service for the Junebug Log Observer.
+
+        :param str worker_id: ID of the worker to observe logs for.
+        :param str path: Path to place the log files.
+        :param int rotate: Size (in bytes) before rotating log file.
+        :param int max_files:
+            Maximum amount of log files before old log files
+            start to get deleted.
+        :param logger:
+            logger to add observer to. Defaults to
+            twisted.python.log.theLogPublisher
+        :type logger: :class:`twisted.python.log.LogPublisher`
+        '''
+        self.setName('Junebug Worker Logger')
+        self.logger = logger if logger is not None else log.theLogPublisher
+        self.worker_id = worker_id
+        self.path = path
+        self.rotate = rotate
+        self.max_files = max_files
+
+    def startService(self):
+        self.logfile = LogFile(
+            self.worker_id, self.path, rotateLength=self.rotate,
+            maxRotatedFiles=self.max_files)
+        self.log_observer = JunebugLogObserver(self.logfile, self.worker_id)
+        self.logger.addObserver(self.log_observer)
+        return super(JunebugLoggerService, self).startService()
+
+    def stopService(self):
+        if self.running:
+            self.logger.removeObserver(self.log_observer)
+            self.logfile.close()
+        return super(JunebugLoggerService, self).stopService()
+
+    def registered(self):
+        return (
+            getattr(self, 'log_observer', None) and
+            self.log_observer in self.logger.observers)
```

Comment:
A common trick is to set `log_observer = None` on the class itself, so that if it hasn't been defined on the instance it can still be accessed.